### PR TITLE
Change type sigs of various internal methods to use size_t and mp_obj_type_t*

### PR DESCRIPTION
**Pull request by dpgeorge**
Now with the option to have 64-bit NaN boxing there is a real distinction between size_t and mp_uint_t.  Eg, on 32-bit arch with 64-bin NaN boxing, size_t is 32-bits while mp_uint_t can be 64-bits wide.  And it's inefficient (in RAM, code size and speed) to unnecessarily use mp_uint_t.

There are three patches in this PR.  Two of them change mp_uint_t to size_t when counting number of arguments (ie number of positional or number of keyword args).  It makes sense to use size_t in such a case because you can't have more arguments than the size of the address space (which is what size_t is for).

The third patch changes the type of the first argument of type.make_new from mp_obj_t to `mp_obj_type_t*`.  This makes sense because in all places where this argument is actually used, it's anyway cast to the latter pointer type.  For object representations where casting like this is a non-trivial operation (eg 64-bit NaN boxing) it makes sense to just pass around the pointer (`mp_obj_type_t*`) from the very beginning.

In summary: for all existing ports there is pretty much no change in generated code, since all types stay the same size and no extra conversion ops are needed.  But for 64-bit NaN boxing it makes a difference.  And anyway I think it's a much cleaner coding style to use these types correctly :)

Reason for the PR/review is to see if there are any large objections, or suggestions for improvement.  And also to advertise such a change, since it touches many files.

### CODE REVIEW

**Comment by dpgeorge** [comment]
If there are no objections or comments then I'll merge this.

**Comment by pfalcon** [comment]
> If there are no objections or comments then I'll merge this.

Sure!

**Comment by dpgeorge** [comment]
Thanks!  Merged.
